File: 6.controllers-services-repositories-middlewares-and-request-context/code/typescript/a_note_on_nodejs_vs_go_python_2.ts
import { Request, Response } from 'express';

const bookService = { listBooks: async (sort: string) => [] }; // sketch

export async function listBooksHandler(req: Request, res: Response) {
    let sort = req.query.sort as string;

    // VALIDATION: if present, must be an allowed value.
    if (sort && sort !== "name" && sort !== "date") {
        res.status(400).json({ error: "sort must be 'name' or 'date'" });
        return;
    }

    // TRANSFORMATION: optional param -> inject a default.
    if (!sort) {
        sort = "date";
    }

    try {
        const books = await bookService.listBooks(sort); // delegate
        res.status(200).json(books); // array of books
    } catch (e) {
        res.status(500).json({ error: "could not fetch books" });
    }
}

File: 6.controllers-services-repositories-middlewares-and-request-context/code/rust/a_note_on_nodejs_vs_go_python_2.rs
use actix_web::{web, HttpResponse, Responder};
use serde::Deserialize;

#[derive(Deserialize)]
struct QueryInfo {
    sort: Option<String>,
}

// Sketch
struct BookService;
impl BookService {
    async fn list_books(&self, sort: &str) -> Result<Vec<String>, ()> { Ok(vec![]) }
}

pub async fn list_books_handler(info: web::Query<QueryInfo>, service: web::Data<BookService>) -> impl Responder {
    let mut sort = info.sort.clone().unwrap_or_default();

    // VALIDATION: if present, it must be one of the allowed values.
    if !sort.is_empty() && sort != "name" && sort != "date" {
        return HttpResponse::BadRequest().body("sort must be 'name' or 'date'");
    }

    // TRANSFORMATION: query params are optional -> inject a default.
    if sort.is_empty() {
        sort = "date".to_string(); // downstream layers never see an empty value
    }

    match service.list_books(&sort).await {
        Ok(books) => HttpResponse::Ok().json(books),
        Err(_) => HttpResponse::InternalServerError().body("could not fetch books"),
    }
}
